query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::rc::Rc;
+
+use mentat_core::{
+    Schema,
+    TypedValue,
+    ValueType,
+};
+
+use mentat_query::{
+    Binding,
+    FnArg,
+    NonIntegerConstant,
+    SrcVar,
+    Variable,
+    WhereFn,
+};
+
+use clauses::{
+    ConjoiningClauses,
+    PushComputed,
+};
+
+use errors::{
+    BindingError,
+    ErrorKind,
+    Result,
+};
+
+use types::{
+    ComputedTable,
+    SourceAlias,
+    VariableColumn,
+};
+
+trait FnArgInto<F> {
+    fn into_typed_value(self, lookup: F) -> Result<TypedValue>;
+    fn into_column_vector(self, lookup: F) -> Result<Vec<TypedValue>>;
+
+    fn into_unit_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+    fn into_row_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+    fn into_column_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+    fn into_block_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>>;
+}
+
+impl<F> FnArgInto<F> for FnArg where F: Fn(Variable) -> Result<TypedValue> {
+    fn into_typed_value(self, lookup: F) -> Result<TypedValue> {
+        match self {
+            FnArg::EntidOrInteger(x) => Ok(TypedValue::Long(x)),
```

Comment:
Okay, so if you `[(ground 1) ?x]` it's always a `Long`. That's unambiguous, but it's not really correct. That makes these two queries return different results:

```
[?find ?x :where [?x _ 1]]
[?find ?x :where [(ground 1) ?y] [?x _ ?y]]
```